File: lunabot_control/src/lunabot_control/nav2_localclient.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from nav2_msgs.action import FollowPath
import threading

class nav2_localclient(Node):
    def __init__(self):
        super().__init__("nav2_localclient_node")
        self._action_client = ActionClient(self, FollowPath, 'follow_path')
        self.create_timer(0.5, self.send_goal)
        
        
    def send_goal(self):
        goal_msg = FollowPath.Goal()
        goal_msg.path.header.frame_id = 'map'
        goal_msg.path.header.stamp = self.get_clock().now().to_msg()
        goal_msg.controller_id = 'FollowPath'
        goal_msg.goal_checker_id = 'simple_goal_checker' 
        
        self._action_client.wait_for_server()
        future = self._action_client.send_goal_async(
            goal_msg, feedback_callback=self.feedback_callback)
        future.add_done_callback(self.done_callback)
        return future
        
    def feedback_callback(self, feedback_msg):
        self.get_logger().info(f'Received feedback: {feedback_msg.feedback}')
    
    def done_callback(self, future):
        self.get_logger().info('Done!')
        

def main():
    rclpy.init()
    action_client = nav2_localclient()
    rclpy.spin(action_client)

File: lunabot_control/src/lunabot_control/nav2_localclient.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "nav2_msgs/action/compute_path_to_pose.hpp"
#include "nav2_msgs/action/follow_path.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "nav_msgs/msg/path.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include <chrono>
#include <rclcpp_action/client.hpp>

using PoseStampedMsg = geometry_msgs::msg::PoseStamped;
using PathMsg = nav_msgs::msg::Path;
using OdometryMsg = nav_msgs::msg::Odometry;
using ComputePathToPose = nav2_msgs::action::ComputePathToPose;

class Nav2LocalClient : public rclcpp::Node {
  rclcpp::TimerBase::SharedPtr timer;
  rclcpp_action::Client<ComputePathToPose>::SharedPtr action;
  rclcpp::Subscription<PoseStampedMsg>::SharedPtr goal_sub;
  rclcpp::Subscription<OdometryMsg>::SharedPtr odom_sub;
  rclcpp::Publisher<PathMsg>::SharedPtr path_pub;

  PathMsg path;
  OdometryMsg odom;

  public:
    Nav2LocalClient() : rclcpp::Node("nav2_localclient_node") {
      odom_sub = create_subscription<OdometryMsg>("/rtabmap/odom", 10, [this] (OdometryMsg value) {
          this->odom = value;
      });
      path_sub = create_subscription<PathMsg>("/test_path", 10, [this] (PathMsg value){
        this->path = value;
      });
     
      action = rclcpp_action::create_client<ComputePathToPose>(this, "/compute_path_to_pose");
      timer = create_wall_timer(std::chrono::milliseconds(500), std::bind(&Nav2LocalCLient::plan_path, this));
    }

  private:
    void plan_path() {
      if (!action->wait_for_action_server()) {
        RCLCPP_WARN(get_logger(), "Action server not ready yet");
        return;
      }

      auto goal = ComputePathToPose::Goal();
      goal.goal = this->goal;
      goal.start.pose = odom.pose.pose; 
      goal.start.header = odom.header; 
      goal.use_start = true;
      goal.planner_id = "GridBased";

      auto options = rclcpp_action::Client<ComputePathToPose>::SendGoalOptions();
      options.result_callback = [this] (rclcpp_action::ClientGoalHandle<ComputePathToPose>::WrappedResult result) {
        if (result.code != rclcpp_action::ResultCode::SUCCEEDED) {
          RCLCPP_WARN(get_logger(), "Failed to compute pose");
        } else {
          this->path_pub->publish(result.result->path);
        }
      };

      action->async_send_goal(goal, options);
    }
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Nav2LocalClient>());
  rclcpp::shutdown();
  return 0;
}